sdks/wasm/Mono.WebAssembly.DebuggerProxy/MonoProxy.cs

```diff
@@ -611,140 +649,116 @@ async Task<DebugStore> LoadStore (SessionId sessionId, CancellationToken token)
 
-			if (!Breakpoint.TryParseId (bpid, out var the_id))
-				return false;
-
 			var context = GetContext (msg_id);
-			var bp = context.Breakpoints.FirstOrDefault (b => b.LocalId == the_id);
-			if (bp == null) {
-				Log ("info", $"Could not find dotnet bp with id {the_id}");
+			if (!context.BreakpointRequests.TryGetValue (bpid, out var breakpointRequest))
 				return false;
-			}
-
-			context.Breakpoints.Remove (bp);
-			//FIXME verify result (and log?)
-			var res = await RemoveBreakpoint (msg_id, bp, token);
 
-			return true;
-		}
-
-
-		async Task<Result> RemoveBreakpoint (SessionId sessionId, Breakpoint bp, CancellationToken token)
-		{
-			var res = await SendMonoCommand (sessionId, MonoCommands.RemoveBreakpoint (bp.RemoteId), token);
-			var ret_code = res.Value? ["result"]? ["value"]?.Value<int> ();
+			foreach (var bp in breakpointRequest.Locations) {
+				var res = await SendMonoCommand (msg_id, MonoCommands.RemoveBreakpoint (bp.RemoteId), token);
+				var ret_code = res.Value? ["result"]? ["value"]?.Value<int> ();
 
-			if (ret_code.HasValue) {
-				bp.RemoteId = -1;
-				bp.State = BreakpointState.Disabled;
+				if (ret_code.HasValue) {
+					bp.RemoteId = -1;
+					bp.State = BreakpointState.Disabled;
+				}
 			}
-
-			return res;
+			breakpointRequest.Locations.Clear ();
+			return false;
 		}
 
-		async Task<bool> SetBreakpoint (MessageId msg_id, BreakpointRequest req, CancellationToken token)
+		async Task SetBreakpoint (SessionId sessionId, DebugStore store, BreakpointRequest req, bool send, CancellationToken token)
 		{
-			var context = GetContext (msg_id);
-			var bp_loc = context.Store.FindBestBreakpoint (req);
-			Log ("info", $"BP request for '{req}' runtime ready {context.RuntimeReady} location '{bp_loc}'");
-			if (bp_loc == null) {
-				Log ("verbose", $"Could not resolve breakpoint request: {req}");
-				return false;
+			var context = GetContext (sessionId);
+			if (req.Locations.Any ()) {
+				Log ("debug", $"locations already loaded for {req.Id}");
+				return;
 			}
 
-			Breakpoint bp = null;
-			if (!context.RuntimeReady) {
-				bp = new Breakpoint (bp_loc, context.NextBreakpointId (), BreakpointState.Pending);
-			} else {
-				bp = new Breakpoint (bp_loc, context.NextBreakpointId (), BreakpointState.Disabled);
-
-				var res = await EnableBreakpoint (msg_id, bp, token);
-				var ret_code = res.Value? ["result"]? ["value"]?.Value<int> ();
+			var locations = store.FindBestBreakpoint (req).ToList ();
+			Log ("info", $"BP request for '{req}' runtime ready {context.RuntimeReady} location '{locations.FirstOrDefault()}'");
 
-				//if we fail we just buble that to the IDE (and let it panic over it)
-				if (!ret_code.HasValue) {
-					return false;
-				}
-			}
+			var breakpoints = new List<Breakpoint> ();
+			foreach (var loc in locations) {
+				var bp = new Breakpoint (req.Id, loc, BreakpointState.Disabled);
+				await EnableBreakpoint (sessionId, bp, token);
```

Comment:
`EnableBreakpoint` can fail, we should handle that here.

Reply:
It is simple enough to check the breakpoint state after the call but when multiple debuggers are running I've seen the call fail because there is already a bp active at that location.  I think we treat that as a failure right now and I'd rather it show success there.  I'll look into the code.

Reply:
Maybe we can change `EnableBreakpoint` to take `loc`+`sessionId` as an argument, and return `Task<Breakpoint>`? The returned `bp` would be in `Active` or `Disabled` state depending on the result of `MonoCommands.SetBreakpoint`. It would be easier to understand, IMHO.

Reply:
It looks like multiple bps are only an issue when looking up the and won't cause a failure here.